extensions-contrib/scan-query/src/main/java/io/druid/query/scan/ScanQueryEngine.java

```diff
@@ -0,0 +1,181 @@
+                          dimSelectors.put(dim.getOutputName(), dimSelector);
+                        }
+
+                        final Map<String, ObjectColumnSelector> metSelectors = Maps.newHashMap();
+                        for (String metric : metrics) {
+                          final ObjectColumnSelector metricSelector = cursor.makeObjectColumnSelector(metric);
+                          metSelectors.put(metric, metricSelector);
+                        }
+                        final int batchSize = query.getBatchSize();
+                        final int limit = query.getLimit();
+                        return new Iterator<ScanResultValue>()
+                        {
+                          private int offset = 0;
+
+                          @Override
+                          public boolean hasNext()
+                          {
+                            return !cursor.isDone() && offset < limit;
+                          }
+
+                          @Override
+                          public ScanResultValue next()
+                          {
+                            int lastOffset = offset;
+                            Object events = null;
+                            String resultFormat = query.getResultFormat();
+                            if ("valueVector".equals(resultFormat)) {
+                              events = rowsToValueVector();
+                            } else {
+                              events = rowsToList();
+                            }
+                            return new ScanResultValue(segmentId, lastOffset, events);
+                          }
+
+                          @Override
+                          public void remove()
+                          {
+                            throw new UnsupportedOperationException();
+                          }
+
+                          private Object rowsToList() {
+                            int i = 0;
+                            List<Map<String, Object>> events = Lists.newArrayListWithCapacity(batchSize);
+                            for (; !cursor.isDone()
+                                   && i < batchSize
+                                   && offset < limit; cursor.advance(), i++, offset++) {
+                              final Map<String, Object> theEvent = SelectQueryEngine.singleEvent(
+                                  ScanResultValue.timestampKey,
+                                  timestampColumnSelector,
+                                  dimSelectors,
+                                  metSelectors
+                              );
+                              events.add(theEvent);
+                            }
+                            return events;
+                          }
+
+                          private Object rowsToValueVector() {
+                            // only support list now, we can support ValueVector or Arrow in future
```

Comment:
UnsupportedOperationException is probably safer, otherwise callers will get things they don't expect